Data/Atomic.hs

```diff
@@ -1,4 +1,9 @@
-{-# LANGUAGE BangPatterns, ForeignFunctionInterface #-}
+{-# LANGUAGE BangPatterns
+           , CPP
+           , ForeignFunctionInterface
+           , MagicHash
+           , UnboxedTuples
+           #-}
 -- | An atomic integer value. All operations are thread safe.
 module Data.Atomic
```

Comment:
I think this module can be swapped out for a thin wrapper over [`atomic-counter`](https://hackage.haskell.org/package/atomic-counter).

Reply:
I've never benchmarked it. Looks like it uses `foreign import prim` instead of using the atomic primops.

Reply:
> By default CMM will be enabled which should save one word of overhead because there would be no array any more hence no need to store trivial size.

Ah that's clever.

Reply:
`foreign import prim` is different from `foreign import ccall`; it's basically how GHC primops are implemented. For sufficiently new GHCs `atomic-counter` takes a shortcut and employs underlying CMM calls instead of GHC primops.

That said, I do not mean to derail your work here by a further digression, I can see that it's been long time already ;)

Reply:
I found the performance to be essentially identical.